### PR TITLE
Development doc

### PR DESCRIPTION
**Pull request by eguzki**

### CODE REVIEW

File: doc/development.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+
+### Inspecting an element
+
+```
+ngx.log(ngx.DEBUG, require("inspect")(object))
+```
+
+### Liquid-lua
+
+Clone [liquid-lua](https://github.com/3scale/liquid-lua) repo.
```
Comment (dfennessy) [inline doc/development.md:11]:
```suggestion
Clone the [liquid-lua](https://github.com/3scale/liquid-lua) repository.
```

File: doc/integration-tests.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
+## Integration testing framework
+
+Test framework used for integration testing in APIcast:
```
Comment (dfennessy) [inline doc/integration-tests.md:3]:
```suggestion
Test the framework used for integration testing in APIcast: 
```

Comment (dfennessy) [inline doc/integration-tests.md:3]:
```suggestion
The test framework used for integration testing in APIcast: 
```

File: doc/integration-tests.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
+## Integration testing framework
+
+Test framework used for integration testing in APIcast: 
+[openresty/test-nginx](https://github.com/openresty/test-nginx): 
+Data-driven test scaffold for Nginx C module and OpenResty Lua library development.
```
Comment (dfennessy) [inline doc/integration-tests.md:5]:
```suggestion
Data-driven test scaffold for NGINX C module and OpenResty Lua library development.
```

File: doc/development.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+Run the following:
+
+```
+docker build --tag lua_test .
+docker run --name luatest -d lua_test sleep infinity
+docker exec -it luatest /bin/bash
+$ vi test01.lua
+[paste one of the examples from https://github.com/3scale/liquid-lua]
+$ lua test01.lua
+```
```
Comment (dfennessy) [inline doc/development.md:45]: Suggest numbering the above since it is a procedure.

File: doc/integration-tests.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+Documentation on this framework: 
+[Automated Testing · Programming OpenResty (gitbooks.io)](https://openresty.gitbooks.io/programming-openresty/content/testing/index.html).
+
+APIcast specific extension/fork of this framework: [3scale/Test-APIcast](https://github.com/3scale/Test-APIcast):
+APIcast testing framework using `Test::Nginx`.
+
+### Run integration tests
+
+Using Docker you just need to run:
```
Comment (dfennessy) [inline doc/integration-tests.md:15]:
```suggestion
Using Docker, run the following command:
```

File: doc/integration-tests.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+```shell
+make development
+```
+
+That will create a Docker container and run bash inside it. The project's source
```
Comment (dfennessy) [inline doc/integration-tests.md:20]:
```suggestion
That creates a Docker container that runs bash inside it. The project's source
```

File: doc/integration-tests.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+code will be available in the container and sync'ed with your local `apicast`
```
Comment (dfennessy) [inline doc/integration-tests.md:21]:
```suggestion
code is then available in the container and synchronized with your local `apicast`
```

File: doc/integration-tests.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+directory, so you can edit files in your preferred environment and still be able
+to run whatever you need inside the Docker container.
+
+To install the dependencies inside the container run:
+```shell
+make dependencies
+```
+
+To run the integration tests inside the container:
```
Comment (dfennessy) [inline doc/integration-tests.md:30]:
```suggestion
For integration tests inside the container, run:
```

File: doc/integration-tests.md [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+```shell
+make development
+```
+
+That will create a Docker container and run bash inside it. The project's source
+code will be available in the container and sync'ed with your local `apicast`
+directory, so you can edit files in your preferred environment and still be able
+to run whatever you need inside the Docker container.
+
+To install the dependencies inside the container run:
```
Comment (dfennessy) [inline doc/integration-tests.md:25]:
```suggestion
To install the dependencies inside the container, run:
```

File: doc/integration-tests.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+```shell
+make dependencies
+```
+
+To run the integration tests inside the container:
+```shell
+make prove
+```
+
+### Useful tips/config & commands
```
Comment (dfennessy) [inline doc/integration-tests.md:35]:
```suggestion
### Useful tips/configuration & commands
```

File: doc/integration-tests.md [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+
+To enable debug logs when running the integration tests set `DEBUG=1` before executing the tests.
+This will print all the logs in `stderr` which is configured in: `$Test::Nginx::Util::ErrLogFile`.
+
+To run a single file of tests
```
Comment (dfennessy) [inline doc/integration-tests.md:40]:
```suggestion
Type the following command to run a single file of tests:
```

File: doc/integration-tests.md [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+the [TEST_NGINX_RANDOMIZE](https://metacpan.org/pod/Test::Nginx::Socket#TEST_NGINX_RANDOMIZE)
+env var to `0`.
+
+Some integration tests are run `n` times where `n` is defined at the top of the test file in the
+function [repeat_each](https://metacpan.org/pod/Test::Nginx::Socket#repeat_each).
+It can be very useful to switch this to 1 when debugging failed tests.
+
+Where env vars such as
+[TEST_NGINX_SERVER_PORT](https://metacpan.org/pod/Test::Nginx::Socket#TEST_NGINX_SERVER_PORT)
+are used in the test, it’s necessary to declare `env` at the top of that section in the test
```
Comment (dfennessy) [inline doc/integration-tests.md:64]:
```suggestion
are used in the test, it is necessary to declare `env` at the top of that section in the test
```

File: doc/integration-tests.md [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+```
+
+For any env vars that include a random port such as `TEST_NGINX_RANDOM_PORT`
+an additional declaration is required; `random_port` only once in the first section of the test
+where that env var is declared. After that `env` is sufficient in any subsequent sections in the
+test to propagate the variable.
+
+#### Integration testing with an HTTP(S) proxy
+
+Mocking of components, configuration & environment in a reusable way is stored in the
```
Comment (dfennessy) [inline doc/integration-tests.md:85]:
```suggestion
Mocking of components, configuration, and environment in a reusable way is stored in the
```

File: doc/integration-tests.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+[t/fixtures](https://github.com/3scale/APIcast/tree/master/t/fixtures) directory.
+In [here](https://github.com/3scale/APIcast/blob/master/t/fixtures/proxy.lua) is the specific Lua
+code for the *http forward proxy* (it’s really just an Nginx reverse proxy).
```
Comment (dfennessy) [inline doc/integration-tests.md:88]:
```suggestion
code for the *http forward proxy*, which is really just an NGINX reverse proxy.
```

File: doc/integration-tests.md [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+The logs are forwarded to `stdout` just like the APIcast & test logs.
```
Comment (dfennessy) [inline doc/integration-tests.md:89]:
```suggestion
The logs are forwarded to `stdout` just like the APIcast and test logs.
```

File: doc/integration-tests.md [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+The Nginx configuration and Perl code to start up the proxy server is
```
Comment (dfennessy) [inline doc/integration-tests.md:90]:
```suggestion
The NGINX configuration and Perl code to start up the proxy server is
```

File: doc/integration-tests.md [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+[here](https://github.com/3scale/APIcast/blob/master/t/http_proxy.pl).
+You can see that the
+[same proxy server](https://github.com/3scale/APIcast/blob/master/t/http_proxy.pl#L5-L8) is used
+for both `HTTP_PROXY` & `HTTPS_PROXY` settings.
```
Comment (dfennessy) [inline doc/integration-tests.md:94]:
```suggestion
for both `HTTP_PROXY` and `HTTPS_PROXY` settings.
```

Comment (dfennessy) [reply doc/integration-tests.md:94]: I suggest using a numbered list for all procedures.

File: README.md [hunk 18]
```diff
@@ -67,7 +66,6 @@ oc new-app -f https://raw.githubusercontent.com/3scale/apicast/master/openshift/
 - Can be deployed in [Openshift](https://www.openshift.com/).
 - Integrates with IDPs like [Keycloak](https://www.keycloak.org) to provide authentication based on [OIDC](https://openid.net/connect/).
 
-
 ## Development
 
 Using Docker you just need to run:
```
Comment (dfennessy) [inline README.md:71]:
```suggestion
Using Docker, run the following command:
```

File: doc/unittests.md [hunk 19]
```diff
@@ -0,0 +1,44 @@
+## Unit testing framework
+
+[busted](https://github.com/Olivine-Labs/busted) is the framework used for the unit tests.
+
+### Run unit tests
+
+Using Docker you just need to run:
```
Comment (dfennessy) [inline doc/unittests.md:7]:
```suggestion
Using Docker, run the following command:
```

File: doc/unittests.md [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+```shell
+make development
+```
+
+That will create a Docker container and run bash inside it. The project's source
```
Comment (dfennessy) [inline doc/unittests.md:12]:
```suggestion
That creates a Docker container that runs bash inside it. The project's source
```

File: doc/unittests.md [hunk 21]
```diff
@@ -0,0 +1,44 @@
...
+code will be available in the container and sync'ed with your local `apicast`
```
Comment (dfennessy) [inline doc/unittests.md:13]:
```suggestion
code is then available in the container and synchronized with your local `apicast`
```

File: doc/unittests.md [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+directory, so you can edit files in your preferred environment and still be able
+to run whatever you need inside the Docker container.
+
+To install the dependencies inside the container run:
```
Comment (dfennessy) [inline doc/unittests.md:17]:
```suggestion
To install the dependencies inside the container, run:
```

File: doc/unittests.md [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+```shell
+make dependencies
+```
+
+To run the unit tests inside the container:
```
Comment (dfennessy) [inline doc/unittests.md:22]:
```suggestion
For unit tests inside the container, run:
```

File: doc/unittests.md [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+```shell
+make busted
+```
+
+To run a single file of tests
```
Comment (dfennessy) [inline doc/unittests.md:27]:
```suggestion
For a single file of tests, run:
```

File: doc/unittests.md [hunk 25]
```diff
@@ -0,0 +1,44 @@
...
+
+```shell
+make busted BUSTED_FILES=path/to/file_spec.lua
+```
+
+To run a specific set of `it` tests, i.e. a contextual description defined
```
Comment (dfennessy) [inline doc/unittests.md:33]:
```suggestion
For a specific set of `it` tests, i.e. a contextual description defined, run:
```

File: doc/unittests.md [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+by the `describe` function, add some unique token, for instance `ONLY`, to the `describe` function.
```
Comment (dfennessy) [inline doc/unittests.md:34]:
```suggestion
by the `describe` function, add some unique token, for instance `ONLY`, to the `describe` function, run:
```

File: doc/unittests.md [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+
+```
+describe("ONLY should be awesome", function()
+```
+
+then, add `BUSTED_ARGS` with `--filter` param
```
Comment (dfennessy) [inline doc/unittests.md:40]:
```suggestion
then, add `BUSTED_ARGS` with `--filter` param:
```

File: doc/unittests.md [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+
+```shell
+make busted BUSTED_FILES=path/to/file_spec.lua BUSTED_ARGS="--filter='ONLY'"
+```
```
Comment (dfennessy) [inline doc/unittests.md:44]: Suggest that all procedures be a numbered list where appropriate.

**Comment by dfennessy** [comment]
@eguzki Comments/suggestions added.

File: doc/development.md (hunk 4 above)
Comment (eguzki) [reply doc/development.md:45]: I think it is pretty clear and the target audience are developers

**Comment by eguzki** [comment]
Thank you very much @dfennessy for the review :medal_sports:

**Comment by eguzki** [comment]
@3scale/apicast ready for review

File: doc/development.md (hunk 4 above)
Comment (dfennessy) [reply doc/development.md:45]: Ok. That makes sense.

**Comment by eguzki** [comment]
https://github.com/orgs/3scale/teams/apicast ready for review

File: doc/development.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (samugi) [reply doc/development.md:45]: This was left pending too some time ago... apologies @eguzki

**Pull request merged**